staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -69,6 +69,11 @@ type Config struct {
 	//       question to this interface as a parameter.  This is probably moot
 	//       now that this functionality appears at a higher level.
 	RetryOnError bool
+
+	// Called whenever the ListAndWatch drops a connection, whether or not
+	// the connection had an error. So that clients can handle errors
+	// and implement backoff.
```

Comment:
Given that the machinery retries, what does "implement backoff" mean here?

Reply:
The retry strategy is currently very naive, and isn't great in cases that require user intervention, see discussion here:
https://github.com/kubernetes/client-go/issues/155
and here:
https://github.com/windmilleng/tilt/issues/2702

In particular, MacOS users were complaining that the exec auth provider was bringing up the macos keychain dialog 1x per second, resulting in thousands of dialogs if they stepped away for a bit.

Reply:
Triggering the exec auth callout should probably be directly rate limited or controlled. I doubt this is the only way to call it inappropriately often. Also, a call to that should be blocking (otherwise you lack something necessary to make a network call at all), so I don't understand how this retry loop could trigger it so many times.

Reply:
It sounds like you're both saying:
1) This fix shouldn't be necessary, and
2) Even if it was necessary, this wouldn't be the right way to fix it.

This makes me think we don't agree on what the diagnosis of the bug is (which is ok! I'm not  convinced I've diagnosed it correctly either!). 

But it seems hard to discuss whether the solution is correct if we don't agree on the diagnosis. What would help us reach agreement on the diagnosis? Are there additional tests that would help?

And then there's the meta question - is it worth spending more time on this PR? If the API is architecturally at odds with the direction this package is going, then I'm happy to cut my sunk costs now, no hard feelings.

Reply:
I feel like this discussion should be out of scope of this PR. It doesn't change anything in that regard, it merely makes it possible to handle the error. Could that be addressed in a separate PR/issue? I think as it is, this PR is a great improvement.

Reply:
Apologies for lack of clarity. I am saying neither of those things.

a) I think there should be a separate rate limiting fix for the auth token getting behavior. Something is weird there that isn't explained by the code here AFAICT. E.g. if the problem is failing to wait for the human to actually click "yes" or "no" then changing this code at best reduces the frequency of that condition.
b) I agree 100% that it makes sense for the retry behavior to be controllable. We have a mostly standardish way of doing that (user constructs and passes a backoff type object).
c) I also agree 100% that it should be discoverable when this thing is not actually connected, and the reason (e.g. most recent error & count).

Where I get off the bus is that I don't think it's good to combine b and c.
* Client-go API changes affect a lot of people and it is worth our time to get them right
* Combining the error reporting and backoff behavior in this way is not consistent with what we do elsewhere in the client
* This code is very tricky; an incorrect delay in this function will make the client much less efficient (the RV is time-sensitive; if re-used rapidly, a watch can be resumed, otherwise a very expensive list has to be done)
* It's very plausible that when the connection is lost, we actually want to mark the cache/informer as "not synced" to prevent controllers from acting with this information

I don't expect a change to perform all of those functions simultaneously, but it needs to not change the API in a way that makes those things harder. Hopefully that makes sense.

In terms of concrete suggestions for this change-- I'm mostly OK with this except for the part where people use this function to do rate limiting. I can imagine either a version of this change where the function is not set up to be delayed (no channel, warning in comment not to do slow processing, etc), or a version where errors are sent through a channel. I think it will be important for the consumer of these channels to also know when we are reconnected, so that they can restart their processing. We can wait for users before adding that, but again the api needs to not rule it out (e.g. OnDisconnection/OnReconnection function pair makes sense).

You can appeal to e.g. @deads2k or @liggitt if you think I'm being too pedantic here, but given the amount of time we've spent making bug fixes to this code, I think I'm being very reasonably cautious.

Reply:
Thank you for the long explanation! It's very helpful. 

One thing I have trouble reconciling - you said "Triggering the exec auth callout should probably be directly rate limited or controlled", but you also said "an incorrect delay in this function will make the client much less efficient". Do you think those are directly opposed constraints? 

I'm asking because I want to make sure that the comments I add address the trade-offs correctly -- would adding more rate limiting to the auth / request layer have bad outcomes for the watch code?

Reply:
> "Triggering the exec auth callout should probably be directly rate limited or controlled", but you also said "an incorrect delay in this function will make the client much less efficient"

The "this" in the latter sentence is talking about the reflector retry loop, not the one needed for the auth token refresh. The need for rate limiting token refreshes is troubling because nothing in our stack is prepared for that to be anything other than instantaneous. We should probably actually start with wrapping that call with https://godoc.org/golang.org/x/sync/singleflight to prevent duplicate calls. That might actually be enough to solve that problem. @mikedanese might have another idea?

Reply:
Can you point me to the MacOS exec auth plugin if it's opensource? Does every call require user interaction? How long do these processes run? How long are the credentials valid for? Clearly, 1000 pop ups is a terrible experience but I would like to have a better idea of what the most desirable behavior would be.

I imagine that this is one of many sites that does not handle this type of blocking interruption well. If we want to support credential plugins that require user interaction, we should do our best to solve these problems at (or closer to) the credential source to limit the impact and complexity we push down to the rest of the client.

Singleflighting and evaluating timeout behavior of the exec credential source sound very beneficial. Most of the mileage we've put on that extension point uses non-interactive login flows so I would be surprised if there weren't rough spots. Also :+1:, thanks for helping prove this use case out.

Reply:
Unfortunately, I don't know : (

The end-user complaining about this wasn't able to repro it reliably. They would take screenshots and logs and relay it to their infrastructure team at their company. Their infrastructure team would complain about it to me, who does not work for that company. It would probably take a while just to explain the question you're asking.

My personal experience with MacOS syscalls is that it can be often non-trivial to figure out when a syscall is going to pop up a dialog box or print something to the terminal.

Sorry, I appreciate that's a frustrating answer.

Reply:
Even (especially?) in the face of not knowing how exactly it works, I think it makes sense to singleflight those calls.

Reply:
A single authenticator *should* already be singleflighted. At least it's supposed to be. I want to check if multiple are getting instantiated in some of the core binaries or if there's something going on.

Reply:
I filed https://github.com/kubernetes/kubernetes/issues/89114